### ISSUE TITLE
Using `request-as-event-emitter` directly

### ISSUE THREAD

**Issue by medikoo** [sindresorhus/got#648]
I got bitten by v9.3 upgrade due to changes in [default options](https://github.com/sindresorhus/got/blob/232e0f365ff4a636e9fc53bda79f4b483998cc50/source/index.js#L5-L46)

Thing is that for our purposes, we construct the request handler on top of [request-as-event-emitter](https://github.com/sindresorhus/got/blob/232e0f365ff4a636e9fc53bda79f4b483998cc50/source/request-as-event-emitter.js) And for that we use [normalize-arguments](https://github.com/sindresorhus/got/blob/232e0f365ff4a636e9fc53bda79f4b483998cc50/source/normalize-arguments.js) directly, however as [default options](https://github.com/sindresorhus/got/blob/232e0f365ff4a636e9fc53bda79f4b483998cc50/source/index.js#L5-L46) are not exposed publicly we were relying on hardcoded version, and with v9.3 it started to crash here: https://github.com/sindresorhus/got/blob/232e0f365ff4a636e9fc53bda79f4b483998cc50/source/normalize-arguments.js#L41 as in our setup we missed [newly added hooks](https://github.com/sindresorhus/got/commit/325409c69fa928380f0ad9b232551a5806dfb58f)

We could avoid that, if [default options](https://github.com/sindresorhus/got/blob/232e0f365ff4a636e9fc53bda79f4b483998cc50/source/index.js#L5-L46) would be exposed publicly (is there an option to do that?) or if handling in normalize-arguments were more gentle to missing options (ideally default options could be part of `normalize-arguments` and eventually exposed from there)

**Comment by sindresorhus** [comment]
> We could avoid that, if default options would be exposed publicly (is there an option to do that?)

You can access the defaults on `got.defaults` or `instance.defaults` if you used `got.create/extend`. I realize now that have haven't properly documented this.

**Comment by sindresorhus** [comment]
> Thing is that for our purposes, we construct the request handler on top of request-as-event-emitter

Would you be able to elaborate on why you need to do that? Maybe there's something we can improve in Got to make your use-case easier to achieve.

**Comment by medikoo** [comment]
> You can access the defaults on got.defaults 

Indeed, I see that now (still in my case I didn't require `got` by main module, but of course I could get around to get that this way)

> Would you be able to elaborate on why you need to do that?

Sure, use cases we have:
- Log following things: request, response status and headers (right after it's received), response body (after complete response is propagated). For first two I could probably rely on `'request'` and `'response'` events, but for last one there appears no clean solution (e.g. retries are handled behind the scene, where I would prefer to have all issued requests logged)
- Preset `options.timeout` for each request automatically (it's not about fixed static value, we run requests in AWS lambda's and want to ensure they timeout before lambda times out)
- Resolve body on basis of content type (e.g. we automatically resolve to JSON if there' response type is `application/json` (no need for `options.json: true`).

Generaly it'll be good to achieve first two use cases without a need of introduction of some decorator that should be required instead - so _aside_ we can hook into general `got` configuration, and then have all requests (made via got required directly) automatically affected.

The last one, is more a feature request, not sure why `got` doesn't do that by default, but maybe there's a good reason

I discovered also few other breaking (on that level) changes between v9.2 and v9.3. So for now, after we patched that, we locked ourselves to `9.3.x`

**Comment by szmarczak** [comment]
> Log following things: request, response status and headers (right after it's received), response body (after complete response is propagated). For first two I could probably rely on 'request' and 'response' events, but for last one there appears no clean solution (e.g. retries are handled behind the scene, where I would prefer to have all issued requests logged)

Use [hooks](https://github.com/sindresorhus/got#hooks) and/or [custom handlers](https://github.com/sindresorhus/got/blob/master/advanced-creation.md#handler). There's a dedicated issue for logging: #559 

```js
const instance = got.extend({
	hooks: {
		beforeRequest: [
			options => {
				console.log('new request:', options);
			}
		],
		afterResponse: [
			response => {
				console.log('statusCode:', response.statusCode);
				console.log('headers:', response.headers);
				console.log('body:', response.body);
				return response;
			}
		]
	}
});
```

> Preset options.timeout for each request automatically (it's not about fixed static value, we run requests in AWS lambda's and want to ensure they timeout before lambda times out)

```js
const instance = (url, options) => got(url, {...options, timeout: computeTimeout()});
```

If it still isn't suitable for you, then you should use [`promise.cancel()`](https://github.com/sindresorhus/got#aborting-the-request).

> Resolve body on basis of content type (e.g. we automatically resolve to JSON if there' response type is application/json (no need for options.json: true).

Very bad idea. But you can still use hooks for that:

```js
const instance = got.extend({
	hooks: {
		afterResponse: [
			response => {
				try {
					const parsed = JSON.parse(response.body);
					response.body = parsed;
				} catch (e) {}

				...

				return response;
			}
		]
	}
});
```

> Generally it'll be good to achieve first two use cases without a need of introduction of some decorator that should be required instead - so that aside we can hook into general got configuration, and then have all requests (made via got required directly) automatically affected.

[`got.create()`](https://github.com/sindresorhus/got/blob/master/advanced-creation.md#gotcreatesettings) / [`got.extend()`](https://github.com/sindresorhus/got#gotextendoptions)

> The last one, is more a feature request, not sure why got doesn't do that by default, but maybe there's a good reason

1. Performance costs.
2. If you query mostly HTML/XML/... you don't want JSON. It's just useless code then.

I could find more reasons but I'm lazy :P

> I discovered also few other breaking (on that level) changes between v9.2 and v9.3. So for now, after we patched that, we locked ourselves to 9.3.x

`request-as-event-emitter` isn't meant to be used publicly. Don't do that. Everything you want to do you don't need that. Use promises or streams.

<b>The [documentation](https://github.com/sindresorhus/got#highlights) doesn't bite :smiley:</b>
You should definitely read it, as it answers all your questions and also contains helpful tips.

**Comment by medikoo** [comment]
>  Use hooks and/or custom handlers.

Mind intention of this issue was to discuss breaking changes between v9.2 and v9.3, and in last comment I was mostly referencing to version v9.2

It's great that `afterResponse` hook was added with v9.3. That helps, still, it doesn't allow to address what I'm after in a way I described (e.g. have also result from `JSON.parse` available)

> const instance = (url, options) => got(url, {...options, timeout: computeTimeout()});

I specifically wrote: _it'll be good to achieve first two use cases without a need of introduction of some decorator that should be required instead - so that aside we can hook into general got configuration, and then have all requests (made via got required directly) automatically affected_

You're hint is nowhere near that.

> Very bad idea. 

That's far from constructive.

> The documentation doesn't bite 😃

I did read it, but it doesn't look that you took effort to actually read with understanding my points, and acknowledge the context.

So I may respond in similar way: __reading with understanding doesn't bite__ 😃

**Comment by sindresorhus** [comment]
> I discovered also few other breaking (on that level) changes between v9.2 and v9.3. So for now, after we patched that, we locked ourselves to 9.3.x

In general, I would always recommend locking down dependencies if you use internal stuff. There's no stability promise for that.

> Resolve body on basis of content type (e.g. we automatically resolve to JSON if there' response type is application/json (no need for options.json: true).

We intentionally don't support that. It's better to be explicit. The request client should be the one deciding what format it expects. This is also how WHATWG Fetch works (`await request.json()`).

> it'll be good to achieve first two use cases without a need of introduction of some decorator that should be required instead - so that aside we can hook into general got configuration, and then have all requests (made via got required directly) automatically affected

So you want a singleton. How about creating the custom Got instance in a file called `got.js` with your customizations and you require that one everywhere instead of `got`? That's how I usually handle it.

**Comment by medikoo** [comment]
> In general, I would always recommend locking down dependencies if you use internal stuff. There's no stability promise for that.

It was not perfectly clear to me that what's in scope of `src` is considered _private_ , but indeed documentation references only `require('got')`, so it's probably more a misunderstanding on my side.

> We intentionally don't support that. It's better to be explicit. The request client should be the one deciding what format it expects. This is also how WHATWG Fetch works (await request.json()).

That's absolutely right to not imply such behavior by default, but when we work with API's we know, having an option to turn on content type based resolution might be handy (anyway I can achieve that now by resorting to hooks introduced with v9.3)

> So you want a singleton. How about creating the custom Got instance in a file called got.js with your customizations and you require that one everywhere instead of got? That's how I usually handle it.

That's what I'm doing now, and it looks with hooks introduced with v9.3 I should be able to achieve what I want without requiring internals directly, so that fixes.

Above I noted, that it might be handy to mock `got` aside (so all further modules that require `got` are affected), and it looks it'll also be doable by having access to default options. Still after giving it a second thought, I'd rather not pursue that, as that'll be less transparent for other developers in the project.
